src/resolvers/dns.rs

```diff
@@ -214,18 +219,35 @@ impl DnsResolverWorker {
         // whether they responded in time or not.
         let first_result = Arc::new(Mutex::new(None));
 
+        // If we get a `NoRecordsFound` error from a server, we won't consider
+        // the client to have errored, but we also don't want to take that
+        // result as authoritative: if dynamic resolution is in play, we might
+        // have one client that only has records for other DNS servers and other
+        // clients that only have records for backends.
```

Comment:
I'm not sure i follow why this is relevant for dynamic resolution , as opposed to a normal lookup.

It's *always* the case that "some DNS servers may have records that are missing from other DNS servers", independent of whether that backend is for a DNS server or something else, right?

Reply:
> It's _always_ the case that "some DNS servers may have records that are missing from other DNS servers", independent of whether that backend is for a DNS server or something else, right?

I think this is technically true but not really expected to be true in practice much, right? Or I'll claim something that seems nearly equivalent: it's always the case that any single DNS server's records may change any time, including right after we get query results. But this resolver caches results for _some_ period of time. Prior to this PR, it treats a successful response from any server as "the results to cache", so doesn't attempt to handle different servers having different records. (To be clear, I think this is perfectly reasonable!)

After this PR, if we were to just treat "no records" as authoritative and ignore any other outstanding queries, that would be correct if we expected all the DNS servers to be more or less interchangeable. But for dynamic resolution specifically, this expectation is wrong: _by design_ the servers are not interchangeable, even though we still query them all.

Reply:
100% agreed with you, until this specific point

> But for dynamic resolution specifically, this expectation is wrong: _by design_ the servers are not interchangeable, even though we still query them all.

To me, the term "dynamic resolution" means "using DNS to find additional DNS servers".

This is why I'm getting stuck on the use of "dynamic resolution" being a motivator. If we spun up a qorb pool with three hard-coded DNS server addresses, we would have **no dynamic resolution**. However, we could still see different results from different DNS servers (maybe they're getting concurrently populated while we're querying them?). Thus, as you say, they still aren't completely interchangeable, even thought we're statically accessing DNS.

Reply:
I think I'll claim that we're already broken in that case, then? If those three servers return different records for the backend we're querying, we're only taking the first to respond and ignoring the others.

Peeling back the curtain a bit: I made the changes to this function last; without them, the new stuff works fine for all the existing qorb tests, _except_ the dynamic resolution test, which needed these changes. That definitely motivated the comment!

Reply:
This may be a sign of lack of sufficient coverage rather than an argument for it being relevant only to dynamic resolution.

Both "looking up DNS servers for dynamic resolution" and "looking up backends, with or without dynamic resolution" use this `dns_lookup` function.

I'm happy with this behavior you're adding, I just think it applies to **all** lookups, and whether or not it's being used to find "DNS servers" or "any other server" is orthogonal.

Reply:
I think we basically agree on all of this, haha. I'll take a shot at rewording the comment.